src/promql/src/extension_plan/scalar_calculate.rs

```diff
@@ -0,0 +1,447 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use datafusion::common::{DFField, DFSchema, DFSchemaRef, Result as DataFusionResult, Statistics};
+use datafusion::execution::context::TaskContext;
+use datafusion::logical_expr::{LogicalPlan, UserDefinedLogicalNodeCore};
+use datafusion::physical_expr::PhysicalSortExpr;
+use datafusion::physical_plan::metrics::{BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, Distribution, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+use datafusion::prelude::Expr;
+use datatypes::arrow::array::{Array, StringArray};
+use datatypes::arrow::compute::concat_batches;
+use datatypes::arrow::datatypes::{Field, Schema, SchemaRef};
+use datatypes::arrow::record_batch::RecordBatch;
+use futures::{ready, Stream, StreamExt};
+use snafu::{OptionExt, ResultExt};
+
+use crate::error::{ColumnNotFoundSnafu, DataFusionPlanningSnafu, Result};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ScalarCalculate {
+    schema: DFSchemaRef,
+    input: LogicalPlan,
+    project_index: (usize, usize),
+    tag_columns: Vec<String>,
+    exprs: Vec<Expr>,
+}
+
+impl ScalarCalculate {
+    pub fn new(
+        input: LogicalPlan,
+        time_index_column: &str,
+        tag_colunms: &[String],
+        field_column: &str,
+    ) -> Result<Self> {
+        let input_schema = input.schema();
+        let ts_index = input_schema
+            .index_of_column_by_name(None, time_index_column)
+            .unwrap_or(None)
+            .context(ColumnNotFoundSnafu {
+                col: time_index_column,
+            })?;
+        let val_index = input_schema
+            .index_of_column_by_name(None, field_column)
+            .unwrap_or(None)
+            .context(ColumnNotFoundSnafu { col: field_column })?;
+        let ts_field = input_schema.field(ts_index);
+        let val_field = DFField::new_unqualified(
+            &format!("scalar({})", field_column),
+            input_schema.field(val_index).data_type().clone(),
+            true,
+        );
+        let schema = DFSchema::new_with_metadata(
+            vec![ts_field.clone(), val_field.clone()],
+            input_schema.metadata().clone(),
+        )
+        .context(DataFusionPlanningSnafu)?;
+        Ok(Self {
```

Comment:
Like adding a blank between the result and the above logic.
```suggestion
        
        Ok(Self {
```